File: src/1410-HTML-Entity-Parser/1410.js
var entityParser = function(text) {
    let res = "", n = text.length, i = 0;
    let entity = {
        '&quot': '\"',
        '&apos': "'",
        '&gt': '>',
        '&lt': '<',
        '&frasl': '/', 
        '&amp': '&',
    };
    
    while (i < n) {
        if (text[i] == '&') {
            let t = "";
            while (text[i] != ';') {
                t += text[i++];
            }

            if (t in entity) res += entity[t];
            else res += t + ";";
        } else res += text[i];
        i++;
    }
    return res;
};

File: src/1410-HTML-Entity-Parser/1410.cpp
class Solution {
public:
    string entityParser(string text) {
        stringstream res;
        int n = text.size(), i = 0;
        unordered_map<string, char> entity = {
            {"&quot", '\"'},
            {"&apos", '\''},
            {"&gt", '>'},
            {"&lt", '<'},
            {"&frasl", '/'}, 
            {"&amp", '&'},
        };
        
        while (i < n) {
            if (text[i] == '&') {
                stringstream t;
                while (text[i] != ';') {
                    t << text[i++];
                }
                
                string cur = t.str();
                if (entity.count(cur)) res << entity[cur];
                else res << cur << ";";
            } else res << text[i];
            i++;
        }
        return res.str();
    }
};